### PR TITLE
Access sharingLinkToRedeem from IOdspResolvedUrl.shareLinkInfo

### LINKED ISSUE #8268: Access sharingLinkToRedeem from IOdspResolvedUrl.shareLinkInfo

### PR DESCRIPTION
**Pull request by ssimic2**
This PR should close https://github.com/microsoft/FluidFramework/issues/8268.

Here we consider backward compatibility between newer ODSP drivers and older ODSP resolvers, as hosts may have bundled (older) resolvers while still relying on new (CDN delivered) drivers.

The changes do NOT consider backward compatibility between new ODSP resolvers and old ODSP drivers, for couple of reasons:
1. ODSP resolvers and drivers are commonly sourced from the same package. This is what we do, for example, for all of our E2E regression tests.
2. In the current cases where there is a mismatch (as it may be with OWH), it is the driver that is delivered via service, and resolver that was bundled. I did not see the use cases to justify added work around providing transient N-1 compatibility in this direction.
3. Makes this transition a 2-step process, vs 3-step.

Follow up work: https://github.com/microsoft/FluidFramework/issues/8784

Targeting 0.56 release.

### CODE REVIEW

**Comment by ssimic2** [comment]
Adding you in @agarwal-navin as you were part of this earlier chat: https://teams.microsoft.com/l/message/19:53328301da384b33bcc81111124e3ab4@thread.skype/1637275433228?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=422665a0-7ad3-4d0d-9171-e8881d0397d9&parentMessageId=1637275433228&teamName=FFX%20(Fluid%20Framework%20and%20Experiences)&channelName=Bugs%20%F0%9F%90%9B&createdTime=1637275433228

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts [hunk 1]
```diff
@@ -88,6 +88,12 @@ export async function fetchSnapshotWithRedeem(
     removeEntries: () => Promise<void>,
     enableRedeemFallback?: boolean,
 ): Promise<ISnapshotContents> {
+    // For back. compatibility purposes. To be removed with #8784.
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:91]: Please use the following format:
// back-compat: <some text, including version when added OR can be removed>

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts [hunk 2]
```diff
@@ -88,6 +88,12 @@ export async function fetchSnapshotWithRedeem(
     removeEntries: () => Promise<void>,
     enableRedeemFallback?: boolean,
 ): Promise<ISnapshotContents> {
+    // back-compat: This block to be removed with #8784 when we only consume/consider odsp resolvers that are 0.51+
```
Comment (jatgarg) [inline packages/drivers/odsp-driver/src/fetchSnapshot.ts:91]: Is it 0.51>= or 0.51>. 
Actually the format is //back-compat last-version-to-back-compat(maybe 0.51 in this case) <comment>

**Review by jatgarg — APPROVED** [review]

File: packages/drivers/odsp-driver/src/fetchSnapshot.ts (hunk 2 above)
Comment (ssimic2) [reply packages/drivers/odsp-driver/src/fetchSnapshot.ts]: Clarified versioning dependency.

**Pull request merged**